### PR TITLE
Add StreamIndex Kernel for CSA [Deepseek v4]

### PR DESCRIPTION
**Pull request by octatrifan**
# Description

Adds a fused Pallas TPU kernel (`csa_streamindex_score`) for DeepSeek-V4 Compressed Sparse Attention (CSA) indexer scoring (for the forward pass)
BUGS: b/552514390

### Summary
- Fuses the $Q \cdot K^T$ matmul, ReLU, head weighting, and causal masking into on-chip VMEM to avoid materializing large intermediate tensors in HBM.
- Supports head-major layout $[B, H, S, D]$ and wraps with `@jax.custom_vjp`.
- Gated behind `use_csa_streamindex_kernel: bool = False` in `base.yml` with automatic fallback to standard einsum.
- Wrapped with `@jax.custom_vjp`: executes the fused Pallas kernel on the forward pass, and falls back to reference einsum autodiff (chunked) on the backward pass.

### Implementation
Tiles query tokens in blocks of 128 and compressed keys in blocks of 1024. For each block, it computes `relu(q @ k.T) * weights`, sums across all 64 heads directly in VMEM, applies the causal mask, and writes out the final `[B, S, W]` scores. This avoids allocating the 4D `[B, H, S, W]` scores tensor in HBM entirely.  It does **not include** the top-k at the end in-kernel.

# Tests

- Unit tests in [`tests/unit/csa_streamindex_test.py`](file:///usr/local/google/home/octatrifan/repos/maxtext/tests/unit/csa_streamindex_test.py):
  - Numerical parity vs reference einsum (with and without causal mask).
  - Custom VJP gradient parity.
  - Layer integration and fallback checks.
  - JAXPR graph verification (`pallas_call` vs `dot_general`).

Reproduce:
```bash
pytest tests/unit/csa_streamindex_test.py
```

# Performance
Microbenchmarks on a single TPU, just for the Index Scoring part (not e2e results). Compared against both the unchunked baseline einsum and the orthogonal head-chunking scan proposed in PR #5116 (which is a pure JAX solution to the OOM problem, but which requires more memory and is slower than StreamIndex)

StreamIndex Pallas Kernel vs. Head Chunking (PR #5116) vs. Baseline
Tested on single-chip TPU v5p. $H=64, D=128, W=S/4$:
| Sequence Length ($S$) | Batch ($B$) | Total Tokens | Baseline Einsum | Head Chunking (PR #5116, $c=16$) | StreamIndex Kernel | Kernel Speedup |
|---|---|---|---|---|---|---|
| 16,384 | 2 | 32,768 | 16.31 ms | 8.45 ms | **5.06 ms** | **1.67x vs PR / 3.22x vs Base** |
| 32,768 | 2 | 65,536 | 64.01 ms | 32.02 ms | **19.77 ms** | **1.62x vs PR / 3.24x vs Base** |
| 65,536 | 2 | 131,072 | OOM (274 GB req) | 125.90 ms (20.09 GiB) | **78.42 ms (18.04 GiB)** | **1.61x vs PR** |
| 65,536 | 4 | 262,144 | OOM | 251.84 ms (40.17 GiB) | **156.84 ms (36.08 GiB)** | **1.61x vs PR** |
| 65,536 | 8 | 524,288 | OOM | OOM | 313.53 ms (72.16 GiB) | **StreamIndex Only** |
| 131,072 | 1 | 131,072 | OOM | 247.27 ms (36.09 GiB) | **156.59 ms (34.04 GiB)** | **1.58x vs PR** |
| 131,072 | 2 | 262,144 | OOM | OOM| **313.14 ms (68.08 GiB)** | **StreamIndex Only** |
| 131,072 | 4 | 524,288 | OOM | OOM | OOM | Physical Limit (>96 GB) |
---
* **Speed**: StreamIndex is **1.6x faster** than Head Chunking and **3.2x faster** than baseline einsum across all context lengths.
* **Scalability**: StreamIndex supports **2x larger contexts** on a single TPU v5p, successfully running at $S=65\text{k}, B=8$ and $S=131\text{k}, B=2$ where Head chunking crashes with OOM.
* **Peak Memory**: StreamIndex uses less HBM by keeping intermediate score tensors on-chip in VMEM rather than HBM.

# Checklist

- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/kernels/attention/csa_streamindex.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+  return dq, dk, dw
+
+
+csa_streamindex_score.defvjp(
+    _csa_streamindex_score_fwd, _csa_streamindex_score_bwd
+)
+
+
+def reference_csa_streamindex_score(
+    q: jax.Array,
```
Comment (dandragona) [inline src/maxtext/kernels/attention/csa_streamindex.py:212]: We should depend on the real pure JAX CSA StreamIndex implementation in MaxText so that this implementation doesn't diverge from what the source.

File: src/maxtext/kernels/attention/csa_streamindex.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+  grid = (batch_size, padded_s // block_q, padded_w // block_w)
+
+  in_specs = [
+      pl.BlockSpec((None, num_heads, block_q, head_dim), lambda b, i, j: (b, 0, i, 0)),
+      pl.BlockSpec((None, block_w, head_dim), lambda b, i, j: (b, j, 0)),
+      pl.BlockSpec((None, block_q, num_heads), lambda b, i, j: (b, i, 0)),
+  ]
+  out_specs = pl.BlockSpec((None, block_q, block_w), lambda b, i, j: (b, i, j))
+
+  out = pl.pallas_call(
```
Comment (dandragona) [inline src/maxtext/kernels/attention/csa_streamindex.py:115]: Swap this with an emit_pipeline, then make a standalone test which activates:

from absl import flags

FLAGS = flags.FLAGS
FLAGS.set_default("xla_xprof_register_llo_debug_info", "True")
FLAGS.set_default("xla_enable_custom_call_region_trace", "True")
FLAGS.set_default("xla_enable_transpose_trace", "True")
FLAGS.set_default("xla_xprof_enable_custom_call_tracing", "True")

and get us the xprof.

File: src/maxtext/layers/attention_compressed.py
Comment (dandragona) [inline src/maxtext/layers/attention_compressed.py:903]: (I know this is pre-existing code) This line makes me sad since we are a pretraining library, is there a situation where future_mask is not None?

Maybe we can fold 909:911 into the else block above?

File: src/maxtext/layers/attention_compressed.py [hunk 3]
```diff
@@ -874,20 +875,28 @@ def indexer_compressor_fn(buf_kv, buf_gate):
...
     q = self.rotary_emb(q, position_ids, unsqueeze_dim=1)
-
-    q = q.astype(jnp.float32)
-    compressed_kv = compressed_kv.astype(jnp.float32)
-
-    scores = jnp.einsum("bhsd,bhwd->bhsw", q, compressed_kv)
-    scores = jax.nn.relu(scores) * self.softmax_scale
     weights = self.weights_proj(hidden_states).astype(jnp.float32) * self.weights_scaling
-    index_scores = jnp.einsum("bhsw,bsh->bsw", scores, weights)
+    if self.config.use_csa_streamindex_kernel:
```
Comment (Rohan-Bierneni) [inline src/maxtext/layers/attention_compressed.py]: Lets make sure that this kernel is only run in training mode. It will break both decode and prefill modes. Can you add a check where this is called only when model_mode == TRAIN

File: src/maxtext/kernels/attention/csa_streamindex.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+  w = w_ref[...].astype(jnp.float32).transpose(1, 0)[:, :, None]
+  s_weighted = jnp.sum(s * w, axis=0) * softmax_scale
+
+  # In-VMEM causal future masking
+  if compress_rate > 0:
+    i = pl.program_id(1)
+    j = pl.program_id(2)
+    q_indices = i * block_q + jnp.arange(block_q, dtype=jnp.int32)[:, None]
+    k_indices = (j * block_w + jnp.arange(block_w, dtype=jnp.int32)[None, :]) * compress_rate
+    future_mask = (k_indices + compress_rate) > (q_indices + 1)
```
Comment (Rohan-Bierneni) [inline src/maxtext/kernels/attention/csa_streamindex.py]: This breaks in MaxText whenever packed sequences or document boundaries are used (where position_ids reset to 0 mid-sequence). 

Also, do we need to do masking in the kernel? attention_compressed.py already materializes future_mask in HBM bc [combined_invalid](https://github.com/AI-Hypercomputer/maxtext/blob/1814305e731d31fc9acd4eee7289285869a28df2/src/maxtext/layers/attention_compressed.py#L913) uses it top invalidate top-k outputs.

I think doing masking outside the kernel is fine, just also need to remove the check in https://github.com/AI-Hypercomputer/maxtext/blob/1814305e731d31fc9acd4eee7289285869a28df2/src/maxtext/layers/attention_compressed.py#L910 as well as this code here. 

Actually, have you tested this edge case where S < block_q with the current setup. I think it might fail

File: src/maxtext/kernels/attention/csa_streamindex.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+  """Forward implementation using fused Pallas TPU kernel for head-major [B, H, S, D] q."""
+  batch_size, num_heads, seq_len, head_dim = q.shape
+  _, compressed_len, comp_head_dim = compressed.shape
+  assert comp_head_dim == head_dim, f"{comp_head_dim=} != {head_dim=}"
+  assert weights.shape == (batch_size, seq_len, num_heads), f"{weights.shape=} != {(batch_size, seq_len, num_heads)=}"
+
+  if block_q is None:
+    block_q = 128 if num_heads >= 32 else 256
+  if block_w is None:
+    block_w = 1024 if num_heads >= 32 else 2048
```
Comment (Rohan-Bierneni) [inline src/maxtext/kernels/attention/csa_streamindex.py]: Did you do any tile size tuning to set these specific default tile sizes? For DsV4, the num_heads=64, so

M = 64 * 128 = 8192
N = 1024
s_2d = 8192 * 1024 * 4 = 32MB

While this is fine for tpu limits when you set the scoped_vmem flag, what is the impact if we did something like block_q=64 or 128 & block_w=512 or 256?

File: src/maxtext/kernels/attention/csa_streamindex.py (hunk 2 above)
Comment (Rohan-Bierneni) [reply src/maxtext/kernels/attention/csa_streamindex.py:130]: +1 to having an xprof trace and a percale trace if possible

File: src/maxtext/kernels/attention/csa_streamindex.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+      functools.partial(
+          csa_streamindex_score_kernel,
+          softmax_scale=softmax_scale,
+          compress_rate=compress_rate,
+      ),
+      in_specs=in_specs,
+      out_specs=out_specs,
+      grid=grid,
+      compiler_params=pltpu.CompilerParams(
+          dimension_semantics=("parallel", "parallel", "arbitrary"),
```
Comment (Rohan-Bierneni) [inline src/maxtext/kernels/attention/csa_streamindex.py]: Marking the compressed key window (calling this W) as arbitrary would tell mosaic to process these in sequential order. 

But IIUC every W writes to an independent slice in out_specs so is it necessary to set it as "arbitrary" instead of "parallel"?

For ref: https://docs.jax.dev/en/latest/_autosummary/jax.experimental.pallas.tpu.CompilerParams.html#jax.experimental.pallas.tpu.CompilerParams.dimension_semantics

File: src/maxtext/kernels/attention/csa_streamindex.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+      softmax_scale=softmax_scale,
+      compress_rate=compress_rate,
+      block_q=block_q,
+      block_w=block_w,
+      interpret=interpret,
+  )
+  return out, (q, compressed, weights)
+
+
+def _csa_streamindex_score_bwd(
```
Comment (Rohan-Bierneni) [inline src/maxtext/kernels/attention/csa_streamindex.py:233]: This should not be using the bwd jax baseline from the initial bringup, but the head chunking jax implementation that should cut the peak HBM by a fact of num_head_chunks (set to 8 i believe). 

Then, it would not materialize the full [B, H, S, W] matrix but [B, num_head_chunks, S, W] at a time.

File: tests/unit/csa_streamindex_test.py [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    w = s // 4
+    scale = d**-0.5
+    key1, key2, key3 = jax.random.split(self.key, 3)
+    q = jax.random.normal(key1, (b, h, s, d), dtype=jnp.bfloat16)
+    compressed = jax.random.normal(key2, (b, w, d), dtype=jnp.bfloat16)
+    weights = jax.random.normal(key3, (b, s, h), dtype=jnp.float32)
+
+    fn = jax.jit(
+        lambda q, k, w: csa_streamindex.csa_streamindex_score(
+            q, k, w, softmax_scale=scale, block_q=128, block_w=512, interpret=False
```
Comment (Rohan-Bierneni) [inline tests/unit/csa_streamindex_test.py]: Match this with the default tile sizes set in the code for block_q & block_w.

File: src/maxtext/layers/attention_compressed.py [hunk 9]
```diff
@@ -874,20 +875,28 @@ def indexer_compressor_fn(buf_kv, buf_gate):
...
+      index_scores = csa_streamindex.csa_streamindex_score(
```
Comment (Rohan-Bierneni) [inline src/maxtext/layers/attention_compressed.py]: How come this is not wrapped with jax.shard_map? Don't we need this to enable compatability with the mesh defined. Seems like it was there but removed in a recent commit

**Review by Rohan-Bierneni — COMMENTED** [review]
Thanks for the pallas kernel implementation! I have left some comments PTAL

File: src/maxtext/kernels/attention/csa_streamindex.py (hunk 1 above)
Comment (octatrifan) [reply src/maxtext/kernels/attention/csa_streamindex.py]: Added `csa_indexer_scores_jax()` as the single definition. The layer's non-kernel path and the kernel's VJP both call it now, so there's nothing left to drift.

File: src/maxtext/kernels/attention/csa_streamindex.py (hunk 2 above)
Comment (octatrifan) [reply src/maxtext/kernels/attention/csa_streamindex.py:127]: Skipping for now out of scope for this PR, as discussed with the commenter. XProf added in the buganizer link.

File: src/maxtext/layers/attention_compressed.py (hunk 3 above)
Comment (octatrifan) [reply src/maxtext/layers/attention_compressed.py]: Done. `use_kernel = config.use_csa_streamindex_kernel` and `model_mode == MODEL_MODE_TRAIN`.

File: src/maxtext/kernels/attention/csa_streamindex.py (hunk 4 above)
Comment (octatrifan) [reply src/maxtext/kernels/attention/csa_streamindex.py]: Removed, the kernel was synthesizing positions from grid indices, which is wrong any time positions `≠ arange(S)`. The caller masks unconditionally now, which is free since future_mask is already materialized for the validity check.

Covered by `test_parity_seq_len_smaller_than_block_q`, which sweeps (S,W) = (1,1), (7,3), (64,16), (127,63).

File: src/maxtext/kernels/attention/csa_streamindex.py (hunk 5 above)
Comment (octatrifan) [reply src/maxtext/kernels/attention/csa_streamindex.py]: Swept on v6e (B=1, H=64, D=128), The smaller tiles you suggested are all slower — 64×256 by 1.7×. 256×1024 is 4% faster but doubles the VMEM footprint, and 128×1024 already has little headroom under the 32 MiB scoped limit at H=64. Keeping 128×1024.

File: src/maxtext/kernels/attention/csa_streamindex.py (hunk 6 above)
Comment (octatrifan) [reply src/maxtext/kernels/attention/csa_streamindex.py]: Changed to all-parallel and verified on v6e.

File: src/maxtext/kernels/attention/csa_streamindex.py (hunk 7 above)
Comment (octatrifan) [reply src/maxtext/kernels/attention/csa_streamindex.py:224]: Done, added `lax.scan` over head chunks, which bounds the saved intermediate to [B, chunk, S, W] instead of [B, H, S, W]

File: tests/unit/csa_streamindex_test.py (hunk 8 above)
Comment (octatrifan) [reply tests/unit/csa_streamindex_test.py]: Exported as `DEFAULT_BLOCK_{Q,W}_{LARGE,SMALL}_H`, and the tests reference the constants instead of hardcoding.

File: src/maxtext/layers/attention_compressed.py (hunk 9 above)
Comment (octatrifan) [reply src/maxtext/layers/attention_compressed.py]: Added, batch-only. It raises if the batch isn't divisible by the shard count. Context/sequence parallelism isn't supported and isn't attempted here.

File: src/maxtext/layers/attention_compressed.py
Comment (octatrifan) [reply src/maxtext/layers/attention_compressed.py]: `future_mask` is a local, only assigned in the AR branch (`~is_valid`, from cache validity). Train and prefill always hit the None path so it's dead for pretraining, live for decode, where it marks unpopulated cache slots and so can't come from position_ids.

Folding into the else is equivalent, but the if `compressed_len == 0`: return sits between them, so it'd compute a `[b, s, 0]` mask the early return discards.

**Review by Rohan-Bierneni — APPROVED** [review]
LGTM. Thanks for the changes!

**Review by shuningjin — APPROVED** [review]

**Review by dandragona — APPROVED** [review]
